source/developer/redux.rst

```diff
@@ -0,0 +1,261 @@
+
+.. code-block:: javascript
+  :linenos:
+
+  export const getUsersByUsername = createSelector(
+      getUsers,
+      (users) => {
+          const usersByUsername = {};
+
+          for (const id in users) {
+              if (users.hasOwnProperty(id)) {
+                  const user = users[id];
+                  usersByUsername[user.username] = user;
+              }
+          }
+
+          return usersByUsername;
+      }
+  );
+
+Here we're using the ``getUsers`` selector to feed users into our function that builds a map of users with username as the key.
+
+So far that's pretty straight forward but what if you want to select some data based on an argument but also need to do some computation? That is a little more tricky if you haven't wrapped your head around the purpose of reselect and how createSelector works, so if you haven't watched the developer talk linked above I would strongly suggest it.
+
+To accomplish this we need to create factory function that will create the selector, instead of just creating the selector directly.
+
+.. code-block:: javascript
+  :linenos:
+
+  function getAllFiles(state) {
+      return state.entities.files.files;
+  }
+
+  function getFilesIdsByPosts(state, post) {
+      return state.entities.files.fileIdsByPostId;
+  }
+
+  export function makeGetFilesForPost() {
+      return createSelector(
+          getAllFiles,
+          getFilesIdsForPost,
+          (state, postId) => postId,
+          (allFiles, fileIdsForPost, postId) => {
+              return fileIdsForPost.map((id) => allFiles[id]);
+          }
+      );
+  }
+
+  // Usage by a third party application
+  const getFilesForPost = makeGetFilesForPost();
+  const files = getFilesForPost(state, 'somepostid');
+
+This can look a bit confusing but there is little happening here we haven't seen before. All that we're doing is using three selectors with ``createSelector``, the third selector just happens to be returning an argument so that our final function has access to it. Remember that every selector always takes state in as the first argument.
+
+If you're thinking, I don't get it why can't we just create the selector normally? Then think about how selectors work and remember that if the state changes then the computation happens again. In this case if our ``postId`` changes that counts as a state change since one of our selectors is returning it. This means every time we provide a different ``postId`` to our selector we lose all the benefits of memoization, which is the whole reason for using reselect. So instead we create a new selector for every post id that we want to select for. That may seem a little crazy at first, but think about how componentized React is and it's not that crazy. All you really need to do is use the factory function to create your selector in the constructor or container of your component and use it solely for that component.
```

Comment:
"If you're thinking, I don't get it why can't we just create the selector normally? Then think about" -> `If you're thinking, "I don't get it. Why can't we just create the selector normally?" then think about`

"In this case if our `postId` changes that counts as a state change since one of our selectors is returning it. This means every time" -> "When the `postId` changes, that counts as a state change, so every time" (The sentences kind of serve the same purpose, so I think we could combine them)
"to our selector we lose all the benefits of memoization, which is" -> "to our selector, we lose all the benefits of memoization which is"

"So instead we create" -> "Instead, we create"
"a new selector for every post id that we want to select for." -> "copies of our selector everywhere we know the post id shouldn't change frequently." (Just to be clear that it's not the worst thing if post id changes)

"but think about" -> "but, if you think about"
"is and it's not that crazy" -> "is, it's not that bad" (comma and avoiding repeated word)
"to create your selector" -> "to create an instance of your selector"
"in the constructor or container of your component" -> "for each component"